node-test/src/test/scala/org/bitcoins/node/UpdateBloomFilterTest.scala

```diff
@@ -0,0 +1,190 @@
+package org.bitcoins.node
```

Comment:
So it seems like these test cases pass for me, but i get this error during the run

```
2019-08-01T14:07:22 ERROR [P2P] Error when processing headers message
java.lang.RuntimeException: Failed to add header to chain, header=0e25e9a65507827aab8b2635e368d6c6286b783df0b74b5f5d155e4fade642c8 reason=BadPreviousBlockHash(BlockHeaderImpl(Int32Impl(536870912),DoubleSha256DigestImpl(6735898d2c3d3d27e3851620cffbd70f62d68dcef42b53dfb82967edcf65353c),DoubleSha256DigestImpl(e323879cc69f50a37b66fc357cfc5d1cd4ba465b439b86b9593be18cdf31f0b6),UInt32Impl(1564668460),UInt32Impl(545259519),UInt32Impl(0)))
        at org.bitcoins.chain.blockchain.ChainHandler.$anonfun$processHeader$1(ChainHandler.scala:62)
        at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:307)
        at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:41)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:92)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:85)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:92)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:49)
        at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

Do we know where this is originating from?

Reply:
After adding some more logs and turning up some logging levels, I found this: 

```log
2019-08-02T09:20:40 INFO [P2P] Received headers message with 107 headers
2019-08-02T09:20:40 INFO [P2P] Received headers=[01ce5a015b44c9c22aca13bfab39ca91789e674353c4cf4100f9ab1759fa2f53,7467cbbf9f83171dcb386c8e2da054588efa08c214974446a7c9e164b97435be,4193402cc3c93c4ce409dabb36a0db1a5973fb221efc98791298947ada2ba6eb,0a4e10ed3f137aad51965ebe73e9baddb7d84ca1364965001fed08d9a537f89f,0dfccb6d7c47a5b6d0312fad8d755672f4ddd0f078fc2c88bacc3f8310c322ea,080e8f8218980d4369c778676c7c1a357795707fc0b0e9d0a95c3b478354e086,00ba4cd6733201b09f57b4f3a0ef5c3ffd653cce3dc1b6306764d1751150e562,43cf8bb247ae5e572f5448e87f93e39c59eb565275ba39cf8b816a0eb40365a6,78003b9c6d9d2b08f2edea9b15c0f2f4c1aba16e2daffc044492c92665d74c65,4391a184dac0c73d660ac9c8ec50d0ce7e8a1490a0ff1bd8a6ef23405a0079c0,66f343930f4142381c8e3076e90026b49e011c5ac6454234a88b8a0c8a81920a,649d3a8f8a839ffe55fec9ac520f962878dc7b1ebcce1a6e9d598b83164b3741,3a78e84f0b9e4584a95138b8abb985077c811dea0bd7e399b72c2727f505dbb9,00ef9f52c0c69d7331808454720ab3a3fb15fde07788e855d17423ba3812997f,07ef3aaae16f78462e04ebc881ff91f6015e873e1d0c56871008e809a88bd9fb,541c4aebab20521a16cdb7fe840530f4ffc56d6bc91cdd61c2c5a82a6b6e0fac,19deffafde5e97bdacfb04f4fff0c91e9bbf13b40283b2f7f3c858528559b799,25744eaacc0afcade75ba088b4734f3468666e144a7e1ecda0dcfaa636871bff,03103377a731fa8a8ca6200a0569bb5dcdad9bc1adaeea40d8013b35bac2184e,4ef0ef707edae73835404655d0ae356d5a2db70b1059b64cf8b04994ffb9c729,41457b8d61f698e04c9bb96a110ce81a697f270228238a28772b627a2c18f507,10ed38600dba7b41e47f016548716a0bf3230094409afb7543d50f577381c8ae,194c65bf69345398539472493081cc361ace7c8e895e69d2bfc5b467ec16330e,76060c29eb39ed4d7053bf4c0b0770018241a824f2f538c4625f57d4265670bb,00fd479732d65a9eeb60e283cd4a322087f5ca2a8453fec04109aff153ff8fd8,498700f9c8b2ced4bf8afce8c37ff82b427febba220918a5f34bfd822922b0a8,6350a800ead3d1bc14f76f5de6d243d121fc1c7ead9f9fd3ac461b738cb857c8,20c9eac8fa7bfe86c812cfa60b8e9b5d4a6ab72dc2147bf4caf3af21c1aede06,2ae0531c2c9f61eafbaf9bf9a72cd672b32a48f3a863b3a6b9920c0abfe8f66f,7d46efb1bae819463e18ae733c982230ec264ed415edcfd8073307c90d42d881,71fc0070613e9222ad3a46648b5ec18eed5fe74b32b3131d39e6c2a43e41c2cf,7e4d9d468bb9f0790df8be13a9937f6f474c2a3fb882d17f901ab81ae53a4073,4cb25a919155c68165de5de0258f714a3f80618851898568b62a3c0487761b14,2d9958c76d74b45db977a8f60358623d03636e8c57de32094382ce8171a362f5,7b8f08fcdb32fc67f5bc715c495bd8061873d1880b62bf2866a6698b43ad07b0,18e51124a3b92c65894f52c609deafff7cbabda0af7c0c57e999f886cadd6bb4,547b63bf879a5d3d1c169d4450425860e0676dd4d97f4955e8cdf5fcb39c0409,3ca11476dc8061a002b375be6f6a77daf21481984be92492384dbc501c97fda8,75a8d0fb6f97d73fdb4bd8a28d1c84cab9e340fdc0da24f8ac04b74c3a8cd44c,2dc7bde2c155f83084d3172227b6d51173c8d2a69f9dce3a57efcb953c2f3174,33b54db0d6e0aa0a6bfe2c9a92237d3e87d020ff14e8662b628e4c8301a96d4b,6d15aebf2c3f7f4845828869cadebae9517905429e47502747d28a8f0e12d153,67462f87f40a72d79b6382a4c44f751b9c394dae432757ca6bba110ab1b37ad1,3050b988e663ee0d237747b9e011359bbfa5da85f12ad4288c57998e1144732e,46d0da614227f4c2d347030bf9c2d82e549af76c70e235cf34a95f074d7e54aa,0257a6fd8804893d2f80a07e8797a4b71ace043bd7d8a12f34f59bfab685c951,1e5bbb8414a59fc2c00e6d6f8b394606ca7a8bbf18b349ecd2a163253079b68d,655237436e33e8c1de19c199ac36a5762b69c731cb7ace21b015a3fb35c7cf23,49abbf1b56d9a59e9c86572fbcd382c084b518a05cd9a3d218e9002b1e037e86,59ca4f35b2a4a6562c8933fd2e3f6752d2ab7ac4d23db08b96167d91c2a89ba4,0308fed9ec4844e6f1eaefa4716891988205516e5129ae3feecadf6b0db3e9e7,16e1b1fbe10326633636375d6a16eb497c61a31af4ac2200250e08076ff4ccd0,59d4465c7d900a0018bec6e46693e55f5a8b74ef1ebee97eee87480e5043ee02,3399a9944bcb472883db977242ebcc04c200f855d8f0865eff7357c2a2eddf0b,7ec6dcba54b156ea0bf8710fd6617e9448756bab52958da076299301d495cc2e,0f77f229054b64b881f970a201340d3d855d8df7f511925827a8aabc9ff7769a,61d2bb337b73479e0a90b61e9a928d73552b8edc886cfd1f958ee072a1927ec6,74dc95be78cb57be074de8588ebd2255139eac4df59461b2356dc02d523d7b8b,43aada4632172552d86a3eec7c947a6ad8b8abd2a1b2205acbcf7248aee5f81a,0629e14aacae59234533aa3a528f629b028accd3e8476d5836175af6ee690b71,72335095013fa3d2356adeebec796e056e880f7b723c8f2c0979c0b330684370,63eb66e74a11b1990866e5740823100b17e82fcd613fb9d665644ef14dd857bc,1ae9616365a1a72581dcc8cedee0cdb36fddeb6abbd8b57c693b96c013beb9d1,358a2a88471fb9b9fab7cd68453887759bacfd1af35dc9a9ccceb1a9018bdb37,192432f6dbb80d1454f6d455879ffb5b50ad2a45add6a02f4775abd7b8feeead,3f7bc22661db7cba36d4f988acfa562d1ba59aca57f3d10bec5f86f59d160b8d,7bbaf935ea3930f0396cb9635839c36a766d4285ef4d359f3aa80b285a8297fc,41e36df33422dd6d2b8f076777417ccc01d4f7d2f8ec6f4e0a8abf4a3847b14f,3924179be5a93d2bb0099344f7019d2d4aeee0f47d3ec3bdad4854c3b1942f0e,35e0a5e0fa80895c61a387230637093c5efff678ad192b8ef06cfdb375853a8e,36bb72acb1ad113016075df8071693613301922aa608635fef65da89c92c0257,61918d95b46acd4c705f6f6723a423586f62e9aeaa9fbad2ba2577bb0bedb2fa,08818aedecadbbfbd33162aaf276b39e06ec8fc69f67ccf2258855db94e7c9e7,6a7f8bd7aa4d8742e23a87ab091beec60ec4a195ecdfb71033f2ae15defeb41a,56d39c99678dc4cd63efdc73f700a3db44062803fe58a4295d3cc3453d563d2b,7b99fef84bf2aaddd49faecfc6be6a1797d6bdb8306bef716d51ca63c36d5bea,6fb47f193e85e188e0a7a038910cdc7552ae2fcb61ba36d279e6a5fe6522845b,169b79e9480363f1f611cbff2ed44ec3c933d8e581f7350514d20f95a7dcf633,7e4798c95ebf5ae3e43bd7d8c461e48f6af460eb4263e2f58008da03d35cd8ae,239867cf6dac10db76d9c13b0fc7d74dce598792b35abef63214a28b938fe57d,28e729cb3fcebc332813a824eb961da34b545443e4834617a8832077b96f303b,3c8ca43b769c4b8c797478904e3fcba83c9d45d9dc7170ebd012453e0c2bd77f,721ee77115b8d895ce8b30a98ccd821f13817043a6addeb34dee1a831e48cc55,648e2d55f39d196b02287e0a9d6d341d959300a5d8f47d7835328b0b7efc655f,658208fa45976f42a990d6ba321e32578320b133658ebd89f52799173d713ecd,3fa5d38a4fc5764d9d2a44a0bd9ea17bf35bee7829b46ed12ac366a2d86c4466,791bdaca5161bfe21860db06c4433c2736936578d498b4e9fd692cf7b32fde3d,5aa248e3df0c166eb42d7c3c9669edcf11ecec9188a21a3de957f96a1f468e4d,4d39ddd37f01d3306cbcc1420c14f9f74a0bfae1c89cea80c1bafc5a2c131679,2aa394e7c6e8a75ff8e7487b5ad37614fcb9d686d937595cbd6b46b6e18fbf92,7d507cdef994faa0d6f84e92be4a391a1b01fa7d5d1c4f0a793e878cb7558f0e,62b80a49f8bb70a3053a501982c0dfd23c8947d7eced96d8947c7ea8b421de10,09fdcbd96b21419ea4c469ede1486072b83c1c2ce2cf5324edfbaab7f5754fe8,596637bead07b447f9d6ae680f43eeafd1485eb99b08e98705255a78a4a02136,76a3a598a25a47a0a009e42dfa1bcdbf4849aec7390c797325abc65e2087bcdc,127a737f69eb8fdb961fada1f31b40e67e7f5361666a39f61135fb61c76700f1,1aa55b08fc336068c1eed1381bb17cf87cb7a78f3cd8754367497923d6072265,351afff9c6ddb66d82f03144d674aa36b416caa3ebe089d2c629185426520101,5d7e7b4f2b8a71ed26dfb7becb046676e3fc4244cd3fdc99e2429125ad3b46b7,5721782f4877abfc704e4f7d658ff901405191ec73a2b1d678bf17c06b7c82e1,23b3e3629772dda9fbf6b1ca519951f05b507fc1d421cb848cdda1d3fd141a9d,760f7a67750525be4b3c6cc4156fdebb9a9d8db48022e2bb4eeed07eaaf7c81c,32bf30ad517030c0f1b37f2889a1dfdbdd8176616bebd328b87bfb08c43dca5f,7b55878946ea54f3de4327225527e06cdfb85d9a047ce9b013e0701177a80dec,2118e19fdd06da3652b4c568e0cd267c5daa015b39f3bfca1b983e7cec408b7c,42073d98da1bc70bd86802b6f06b0982fbf2c68d5f58bd0643caebb5cbc35d22,129746ac3eb38fe202d79ddd597c9ce5b254261dea4300e6d724f957cef9d0b9]
2019-08-02T09:20:40 DEBUG [chain-verification] Processing header=01ce5a015b44c9c22aca13bfab39ca91789e674353c4cf4100f9ab1759fa2f53, previousHash=0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206
2019-08-02T09:20:40 DEBUG [chain-verification] Attempting to add new tip=01ce5a015b44c9c22aca13bfab39ca91789e674353c4cf4100f9ab1759fa2f53 with prevhash=0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206 to chain
2019-08-02T09:20:40 DEBUG [chain-verification] getBlockCount result: count=107
2019-08-02T09:20:40 INFO [KEY-HANDLING] Sending 5 BTC to 2MyZE8Yo1aPH14MMgxAGFJj6caMQFFDHE9t at feerate SatoshisPerByte(100 sats)
2019-08-02T09:20:40 DEBUG [chain-verification] Successfully verified=01ce5a015b44c9c22aca13bfab39ca91789e674353c4cf4100f9ab1759fa2f53, connecting to chain
2019-08-02T09:20:40 DEBUG [chain-verification] Connected new header to blockchain, height=1 hash=01ce5a015b44c9c22aca13bfab39ca91789e674353c4cf4100f9ab1759fa2f53
2019-08-02T09:20:40 DEBUG [chain-verification] Processing header=7467cbbf9f83171dcb386c8e2da054588efa08c214974446a7c9e164b97435be, previousHash=01ce5a015b44c9c22aca13bfab39ca91789e674353c4cf4100f9ab1759fa2f53
2019-08-02T09:20:40 DEBUG [chain-verification] Attempting to add new tip=7467cbbf9f83171dcb386c8e2da054588efa08c214974446a7c9e164b97435be with prevhash=01ce5a015b44c9c22aca13bfab39ca91789e674353c4cf4100f9ab1759fa2f53 to chain
2019-08-02T09:20:40 INFO [KEY-HANDLING] Spending UTXOs: 8a54a692299ccdaba617d67015747b631d7a7ed576ddafaabf4be9f5c6c6f902:0
2019-08-02T09:20:40 INFO [KEY-HANDLING] UTXO 0 details: TransactionOutput(2500000000 sats,P2PKHScriptPubKeyImpl(1976a914ba90ac75971a4bc254a3a09b06056703cbb96e0288ac))
2019-08-02T09:20:40 DEBUG [chain-verification] Successfully verified=7467cbbf9f83171dcb386c8e2da054588efa08c214974446a7c9e164b97435be, connecting to chain
2019-08-02T09:20:40 DEBUG [chain-verification] Connected new header to blockchain, height=2 hash=7467cbbf9f83171dcb386c8e2da054588efa08c214974446a7c9e164b97435be
2019-08-02T09:20:40 DEBUG [chain-verification] Processing header=4193402cc3c93c4ce409dabb36a0db1a5973fb221efc98791298947ada2ba6eb, previousHash=7467cbbf9f83171dcb386c8e2da054588efa08c214974446a7c9e164b97435be
2019-08-02T09:20:40 DEBUG [chain-verification] Attempting to add new tip=4193402cc3c93c4ce409dabb36a0db1a5973fb221efc98791298947ada2ba6eb with prevhash=7467cbbf9f83171dcb386c8e2da054588efa08c214974446a7c9e164b97435be to chain
2019-08-02T09:20:40 INFO [KEY-HANDLING] Processing TX from our wallet, transaction=DoubleSha256BDigestBEImpl(fa63c81841b2ffdb757b2ae5c8f93c321c34ca17a77643160370298fdd74c8f5) with confirmations=0
2019-08-02T09:20:40 DEBUG [chain-verification] Successfully verified=4193402cc3c93c4ce409dabb36a0db1a5973fb221efc98791298947ada2ba6eb, connecting to chain
2019-08-02T09:20:40 DEBUG [chain-verification] Connected new header to blockchain, height=3 hash=4193402cc3c93c4ce409dabb36a0db1a5973fb221efc98791298947ada2ba6eb
2019-08-02T09:20:40 DEBUG [chain-verification] Processing header=0a4e10ed3f137aad51965ebe73e9baddb7d84ca1364965001fed08d9a537f89f, previousHash=4193402cc3c93c4ce409dabb36a0db1a5973fb221efc98791298947ada2ba6eb
2019-08-02T09:20:40 DEBUG [chain-verification] Attempting to add new tip=0a4e10ed3f137aad51965ebe73e9baddb7d84ca1364965001fed08d9a537f89f with prevhash=4193402cc3c93c4ce409dabb36a0db1a5973fb221efc98791298947ada2ba6eb to chain
2019-08-02T09:20:40 INFO [KEY-HANDLING] Adding UTXO to wallet: f5c874dd8f297003164376a717ca341c323cf9c8e52a7b75dbffb24118c863fa:1
2019-08-02T09:20:40 INFO [KEY-HANDLING] Successfully inserted UTXO f5c874dd8f297003164376a717ca341c323cf9c8e52a7b75dbffb24118c863fa:1 into DB
2019-08-02T09:20:40 INFO [KEY-HANDLING] Processing of internal transaction=DoubleSha256BDigestBEImpl(fa63c81841b2ffdb757b2ae5c8f93c321c34ca17a77643160370298fdd74c8f5) resulted in changeOutputs=1 and spentUTXOs=1
2019-08-02T09:20:40 INFO [P2P] Sending out inv for tx=DoubleSha256BDigestBEImpl(fa63c81841b2ffdb757b2ae5c8f93c321c34ca17a77643160370298fdd74c8f5)
2019-08-02T09:20:40 INFO [P2P] Updating bloom filter with transaction=DoubleSha256BDigestBEImpl(fa63c81841b2ffdb757b2ae5c8f93c321c34ca17a77643160370298fdd74c8f5)
2019-08-02T09:20:40 DEBUG [chain-verification] Successfully verified=0a4e10ed3f137aad51965ebe73e9baddb7d84ca1364965001fed08d9a537f89f, connecting to chain
2019-08-02T09:20:40 DEBUG [chain-verification] Connected new header to blockchain, height=4 hash=0a4e10ed3f137aad51965ebe73e9baddb7d84ca1364965001fed08d9a537f89f
2019-08-02T09:20:40 DEBUG [chain-verification] Processing header=0dfccb6d7c47a5b6d0312fad8d755672f4ddd0f078fc2c88bacc3f8310c322ea, previousHash=0a4e10ed3f137aad51965ebe73e9baddb7d84ca1364965001fed08d9a537f89f
2019-08-02T09:20:40 DEBUG [chain-verification] Attempting to add new tip=0dfccb6d7c47a5b6d0312fad8d755672f4ddd0f078fc2c88bacc3f8310c322ea with prevhash=0a4e10ed3f137aad51965ebe73e9baddb7d84ca1364965001fed08d9a537f89f to chain
2019-08-02T09:20:40 INFO [P2P] Received headers message with 1 headers
2019-08-02T09:20:40 INFO [P2P] Received headers=[10e367cd4d37fdf590ed3d7beb663ba7207db51f2cfd291e26ccb199542ec6db]
2019-08-02T09:20:40 DEBUG [chain-verification] Processing header=10e367cd4d37fdf590ed3d7beb663ba7207db51f2cfd291e26ccb199542ec6db, previousHash=129746ac3eb38fe202d79ddd597c9ce5b254261dea4300e6d724f957cef9d0b9
2019-08-02T09:20:40 DEBUG [chain-verification] Attempting to add new tip=10e367cd4d37fdf590ed3d7beb663ba7207db51f2cfd291e26ccb199542ec6db with prevhash=129746ac3eb38fe202d79ddd597c9ce5b254261dea4300e6d724f957cef9d0b9 to chain
2019-08-02T09:20:40 WARN [chain-verification] No common ancestor found in the chain to connect header=10e367cd4d37fdf590ed3d7beb663ba7207db51f2cfd291e26ccb199542ec6db
2019-08-02T09:20:40 WARN [chain-verification] Failed to add header to chain, header=10e367cd4d37fdf590ed3d7beb663ba7207db51f2cfd291e26ccb199542ec6db reason=BadPreviousBlockHash(hash=DoubleSha256BDigestBEImpl(10e367cd4d37fdf590ed3d7beb663ba7207db51f2cfd291e26ccb199542ec6db), previous=DoubleSha256BDigestBEImpl(129746ac3eb38fe202d79ddd597c9ce5b254261dea4300e6d724f957cef9d0b9))
2019-08-02T09:20:40 WARN [chain-verification] We don't have competing chainTips. Most recent, valid header=61d66cc5391d59fe4d32d3bc79dbfd04fb840b6cb29d03da640da75e381a8a39
2019-08-02T09:20:40 ERROR [P2P] Error when processing headers message
java.lang.RuntimeException: Failed to add header to chain, header=10e367cd4d37fdf590ed3d7beb663ba7207db51f2cfd291e26ccb199542ec6db reason=BadPreviousBlockHash(hash=DoubleSha256BDigestBEImpl(10e367cd4d37fdf590ed3d7beb663ba7207db51f2cfd291e26ccb199542ec6db), previous=DoubleSha256BDigestBEImpl(129746ac3eb38fe202d79ddd597c9ce5b254261dea4300e6d724f957cef9d0b9))
```

I recommend opening it in a text editor to search in it. What happens: 
1) We receive a `headers` message with 107 headers (hash of last header: 02d79ddd597c9ce5b254261dea4300e6d724f957cef9d0b9)
2) After we've synced 4 of them, we receive a new `headers` message, with 1 header: 10e367cd4d37fdf590ed3d7beb663ba7207db51f2cfd291e26ccb199542ec6db, that references the last header received in step 1)
3) We process the newly arrived header, while we're still churning away at the first 107
4) We find no tip to connect the one from step 2), so we fail

Reply:
Fairly certain that this is the root cause: https://github.com/bitcoin-s/bitcoin-s/pull/655#discussion_r308217009